cartridges/int_algolia/cartridge/scripts/algolia/helper/AlgoliaJobReport.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+
+/**
+* @class AlgoliaJobReport
+* @description Iterator for the changedProducts array of objects structure used in the delta export job
+* @param {string} jobID - the ID or name of the job
+* @param {string} jobType - 'product' or 'category'
+* @constructor
+*/
+const AlgoliaJobReport = function(jobID, jobType) {
+
+    this.jobID = jobID;
+    this.jobType = jobType;
+
+    this.startTime;
+    this.endTime;
+
+    this.processedItems = 0;
+    this.processedItemsToSend = 0;
+    this.siteLocales = 0;
+    this.recordsToSend = 0;
+    this.recordsSent = 0;
+    this.recordsFailed = 0;
+    this.chunksSent = 0;
+    this.chunksFailed = 0;
+
+    this.error = false;
+    this.errorMessage = '';
+}
+
+AlgoliaJobReport.prototype.writeToCustomObject = function() {
+    const CustomObjectMgr = require('dw/object/CustomObjectMgr');
+    const Transaction = require('dw/system/Transaction');
+    const System = require('dw/system/System');
+    const StringUtils = require('dw/util/StringUtils');
+
+    const customObjectID = this.jobID + '__' + StringUtils.formatCalendar(System.getCalendar(), 'yyMMdd-HHmmss'); // returns GMT time
+
+    let that = this; // context changes inside Transaction.wrap()
+
+    try {
+        Transaction.wrap(function() {
+            const algoliaJobReportCO = CustomObjectMgr.createCustomObject('AlgoliaJobReport', customObjectID);
+
+            algoliaJobReportCO.custom.jobID = that.jobID;
+            algoliaJobReportCO.custom.jobType = that.jobType;
+
+            algoliaJobReportCO.custom.startTime = empty(that.startTime) ? null : that.startTime;
+            algoliaJobReportCO.custom.endTime = empty(that.endTime) ? null : that.endTime;
+
+            algoliaJobReportCO.custom.processedItems = that.processedItems;
+            algoliaJobReportCO.custom.processedItemsToSend = that.processedItemsToSend;
+            algoliaJobReportCO.custom.siteLocales = that.siteLocales;
+
+            algoliaJobReportCO.custom.recordsToSend = that.recordsToSend;
+            algoliaJobReportCO.custom.recordsSent = that.recordsSent;
+            algoliaJobReportCO.custom.recordsFailed = that.recordsFailed;
+            algoliaJobReportCO.custom.chunksSent = that.chunksSent;
+            algoliaJobReportCO.custom.chunksFailed = that.chunksFailed;
+
+            algoliaJobReportCO.custom.error = that.error;
+            algoliaJobReportCO.custom.errorMessage = that.errorMessage;
+
+
+        });
+    } catch (e) {
+        require('dw/system/Logger').getLogger('algolia', 'Algolia').error('Failed to write AlgoliaJobReport to custom object: ' + e.message);
+        return false;
+    }
+
+    return true;
+}
+
+AlgoliaJobReport.prototype.formatCustomObject = function(customObject) {
+    const System = require('dw/system/System');
```

Comment:
Yeah, it was initially only used in the other method, but not this one, so didn't want to import it unnecessarily at the top. Now that it's used by both, I'll import it at the top.

Reply:
> to import it unnecessarily at the top

How can it be unnecessarily, all methods in this file will be used by a running job at some point, so all require will be called eventually.

Ok so if I understand correctly, there is no SFCC-related technical reason?
I'm asking because a good practice in Node.js is to put all require at the top level, to identify easily the dependencies: https://eslint.org/docs/latest/rules/global-require
(it's even mandatory now in ES6)

Reply:
While `require()`'d modules remain cached throughout the execution of a single job / request, the two methods in this module are never called in the same request, so it really is unnecessary to require a module at the top which is only used in one of them. `writeToCustomObject` is called from within the jobs, while `formatCustomObject` is called when the Algolia BM module is opened.

That mandatory part in the document refers to `import` and `export` statements, not `require`s (which can be used anywhere in the code). While from a readability standpoint it can be nice to have all `require`'s listed at the top, it would be at the cost of performance unless all of them are used in all methods of the script.
Local `require`'s are better from an optimization standpoint.